Convert documentation to Sphinx

termim opened the pull request:
Please consider this conversion documentation files to Sphinx.
The result could be seen here: http://fio.readthedocs.io/en/sphinx-doc/

axboe commented:
This looks nifty - if we go this route, we can dump fio.1 and just generate it form the HOWTO, correct?

termim commented:
Correct, I've fixed conf.py to generate fio.1 from the HOWTO only.
Just run 'make man' in the doc directory and it will create _build/man/fio.1

axboe commented:
Awesome, this is great. Before we go all in on this and kill off the current fio.1 man page, I need to do a full check between HOWTO and fio.1 to ensure that everything that is in fio.1 is currently in HOWTO. Did you by any chance take a look at that?

termim commented:
There are some options in fio.1 that are missing in the HOWTO.
I could sync them but after NY most probably.

sitsofe commented:
The increased hyperlinking and anchors really make referring to and moving around the converted HOWTO dramatically better! Plus I think maintaining the prose documentation in one place is much better. I've got a few comments/suggestions on the current conversion:

Is it possible to keep the full ReST text of the HOWTO under the filename HOWTO.rst ? Github has limited understanding of ReST and won't follow includes (see https://github.com/termim/fio/blob/sphinx-doc/doc/HOWTO.rst ) and won't specially render ReST file without an extension (see https://github.com/termim/fio/blob/sphinx-doc/HOWTO ). Perhaps the current HOWTO file could be the one turned into a placeholder pointing elsewhere?

Something I noticed after the conversion is that certain quotes have become "smart" which will break copy/paste attempts. For example see http://fio.readthedocs.io/en/sphinx-doc/HOWTO.html#cmdoption-arg-buffer_pattern - `where the string must then be wrapped with “”` and `buffer_pattern=0xdeadface”abcd”-12`. Another example can be seen on http://fio.readthedocs.io/en/sphinx-doc/HOWTO.html#cmdoption-arg-filename : `then you would use filename=”/dev/dsk/foo@3,0:c”. ‘-‘ is a reserved name, meaning stdin or stdout.` (all the quotes in that are smart and the single ones look a bit strange in that context). Similarly an em dash is often seen replacing the two dashes of options. For example in http://fio.readthedocs.io/en/sphinx-doc/HOWTO.html#running-fio there is `would be –iodepth 2 or –iodepth=2. You can also use the command line for giving more than one job entry. For each –name option` and in http://fio.readthedocs.io/en/sphinx-doc/HOWTO.html#terse-output there is `use the –minimal command`. 

There's the occasional capitalisation where I'm not sure it's required (http://fio.readthedocs.io/en/sphinx-doc/HOWTO.html#cmdoption-arg-replay_redirect `major/minor mapping. Replay_redirect causes`). Should all references to options be wrapped in markup?

The ioengine options section looks a bit strange now - http://fio.readthedocs.io/en/sphinx-doc/HOWTO.html#cmdoption-arg-userspace_reap . Perhaps they should be done as subsubsections or as a table?

The "Interpreting the output" table (http://fio.readthedocs.io/en/sphinx-doc/HOWTO.html#interpreting-the-output ) is fantastic!

sitsofe commented:
These commits will probably need rebasing because of doc commits  like 6d500c2e886635861f68cac61039da1f397a7685 and 95a12fbe88afa768926403c5a4ef53346ba2544d ...

termim commented:
On 12/28/16 12:21, Jens Axboe wrote:
>
> Awesome, this is great. Before we go all in on this and kill off the 
> current fio.1 man page, I need to do a full check between HOWTO and 
> fio.1 to ensure that everything that is in fio.1 is currently in 
> HOWTO. Did you by any chance take a look at that?
>
I've synced job file parameters in HOWTO with fio.1. I've tried to 
follow commits to these files and select the latest ones. There are some 
cases when the descriptions where different though.

The information contained in fio.1 is spread in both README and HOWTO. 
To generate equivalent fio.1 automatically we need either reshuffle them 
or combine into one file. I think we could move all detailed information 
to the HOWTO file and leave just generic one in the README. What do you 
think?

axboe commented:
On 01/08/2017 06:22 PM, Mikhail Terekhov wrote:
> On 12/28/16 12:21, Jens Axboe wrote:
>>
>> Awesome, this is great. Before we go all in on this and kill off the
>> current fio.1 man page, I need to do a full check between HOWTO and
>> fio.1 to ensure that everything that is in fio.1 is currently in
>> HOWTO. Did you by any chance take a look at that?
>>
> I've synced job file parameters in HOWTO with fio.1. I've tried to
> follow commits to these files and select the latest ones. There are some
> cases when the descriptions where different though.

Great!

> The information contained in fio.1 is spread in both README and HOWTO.
> To generate equivalent fio.1 automatically we need either reshuffle them
> or combine into one file. I think we could move all detailed information
> to the HOWTO file and leave just generic one in the README. What do you
> think?

I think that's a fair suggestion - leave the command info in the HOWTO, and
keep the README basic, and then we can generate fio.1 etc from the HOWTO.

-- 
Jens Axboe

termim commented:
On 1/8/17 21:09, Jens Axboe wrote:

> I think that's a fair suggestion - leave the command info in the 
> HOWTO, and
> keep the README basic, and then we can generate fio.1 etc from the HOWTO.
I've moved some sections between README and HOWTO. Please take a look.

--
Mikhail Terekhov

termim commented:
On 12/31/16 09:54, Sitsofe Wheeler wrote:

> The increased hyperlinking and anchors really make referring to and 
> moving around the converted HOWTO dramatically better! Plus I think 
> maintaining the prose documentation in one place is much better. I've 
> got a few comments/suggestions on the current conversion:
>
> Is it possible to keep the full ReST text of the HOWTO under the 
> filename HOWTO.rst ? Github has limited understanding of ReST and 
> won't follow includes (see 
> https://github.com/termim/fio/blob/sphinx-doc/doc/HOWTO.rst ) and 
> won't specially render ReST file without an extension (see 
> https://github.com/termim/fio/blob/sphinx-doc/HOWTO ). Perhaps the 
> current HOWTO file could be the one turned into a placeholder pointing 
> elsewhere?
>
It is up to Jens to decide but the problem is that Sphinx doesn't 
support files without extension, so some
file(s) with some extension is(are) needed to include current README, 
HOWTO, etc.

> Something I noticed after the conversion is that certain quotes have 
> become "smart" which will break
>
Thanks, it really helps to have several more eyes to check for 
correctness :)

> copy/paste attempts. For example see 
> http://fio.readthedocs.io/en/sphinx-doc/HOWTO.html#cmdoption-arg-buffer_pattern 
> - |where the string must then be wrapped with “”| and 
> |buffer_pattern=0xdeadface”abcd”-12|.
>
How do you think it should look?

> Another example can be seen on 
> http://fio.readthedocs.io/en/sphinx-doc/HOWTO.html#cmdoption-arg-filename 
> : |then you would use filename=”/dev/dsk/foo@3,0:c”. ‘-‘ is a reserved 
> name, meaning stdin or stdout.| (all the quotes in that are smart and 
> the single ones look a bit strange in that context).
>
Fixed.
>
> Similarly an em dash is often seen replacing the two dashes of 
> options. For example in 
> http://fio.readthedocs.io/en/sphinx-doc/HOWTO.html#running-fio there 
> is |would be –iodepth 2 or –iodepth=2. You can also use the command 
> line for giving more than one job entry. For each –name option| and in 
> http://fio.readthedocs.io/en/sphinx-doc/HOWTO.html#terse-output there 
> is |use the –minimal command|
>
> There's the occasional capitalisation where I'm not sure it's required 
> (http://fio.readthedocs.io/en/sphinx-doc/HOWTO.html#cmdoption-arg-replay_redirect 
> |major/minor mapping. Replay_redirect causes|). Should all references 
> to options be wrapped in markup?
>
Fixed.
>
> The ioengine options section looks a bit strange now - 
> http://fio.readthedocs.io/en/sphinx-doc/HOWTO.html#cmdoption-arg-userspace_reap 
> . Perhaps they should be done as subsubsections or as a table?
>
I'm not sure it is possible to put options definition into table ...

I've reshuffled README and HOWTO a lot, please take another look.

sitsofe commented:
@termim Anything I find now is going to minor - you've done a great job!

(The buffer_pattern=0xdeadface problem has been resolved and looks as I would expect by using non-smart quotes)

Sometimes when writing "fio" the first letter capitalised at the starts of sentences and sometimes it's not. Sometimes fio is written in all caps. Sometimes fio is emboldened. For example: 
http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#source
"fio resides in a git repo[...]"

http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#author
"Fio was written by Jens Axboe[...]"

http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#building
"To build FIO with a cross-compiler:"

http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#platforms
"Fio works on (at least) Linux, Solaris,[...]"

http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption--section
"Tell **fio** to run only the [...]"

Personally like the capitalisation of the first letter at the start of a sentence but otherwise I'd say fio should be all lowercase. Thoughts?

Sometimes paths have formatting sometimes they don't
http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#windows
"[...] dobuild.cmd from the os/windows"
(`os/windows` is a path in this case and `dobuild.cmd` is actually a filename)

http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#documentation :
"[...] and direct your browser to `./doc/output/html/index.html`."

The dash in the following should probably be an em dash (were you getting those by using -- ?):
http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#how-fio-works :
"A job file may contain any number of threads and/or files - the typical contents of the job [...]"

In http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#command-line-options sometimes commands that have a singl sentence end in a full stop and sometimes they don't:
http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption--parse-only
"Parse options only, don’t start any I/O"
http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption--minimal
"Print statistics in a terse, semicolon-delimited format."

Perhaps now we have the ablility to use formatting the text in the brackets should be changed  in http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption--idle-prof ? Other commands seem to use "<option>", "<option>" ...:
```
Report cpu idleness on a system or percpu basis (option=`system,percpu`) or run unit work calibration only (option=`calibrate`).
```

Line wrapping looks strange and "af" should probably be "at" in http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption--trigger-timeout :
"**--trigger-timeout**=t Execute
trigger af this time"

In http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#job-file-format one of the example job files is not formatted with .ini colouring (making it stand out from the rest):
```
filename directive, as in the following example:

; -- start job file including.fio --
[global]
[...]
```

Should there be extra formatting for things that are to do with VARNAME in http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#environment-variables ?

In http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#parameter-types there is this:
```
1 MB: 1000000, 1m, 1000k
1 TiB: 1073741824, 1ti, 1024mi, 1048576ki
1 TB: 1000000000, 1t, 1000m, 1000000k
````

Isn't 1TiB = 1024**4 = 1099511627776 rather than 1073741824? Perhaps the unit for that row were meant to be 1 GiB and 1gi? There might be something similar later too:
```
1 MB: 1000000, 1mi, 1000ki
1 TiB: 1073741824, 1t, 1024m, 1048576k
1 TB: 1000000000, 1ti, 1000mi, 1000000ki
```

Perhaps there should be a back link to the int section of http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#parameter-types in http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-kb_base ?

More smart quotes in unexpected places in http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-directory :
```
Used to place files in a different location than :file:”./”. 
```

In http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-file_service_type perhaps the example should just be verbatim:
"For example, specifying **file_service_type** = *random:8* would"
file_service_type is a self hyperlink but the whole thing is kind of an example so I'd expect the same rendering as `rw=write:4k` in http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-readwrite .

Backslash needs escaping in http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-filename :
"[...] then escape that with a ‘’ character."
(notice how there's nothing between the quotes)

Missing formatting around "rw=randread:8" in http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-rw_sequencer .

The vertical spacing looks weird in http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-iomem :
```
malloc

Use memory from malloc(3) as the buffers. Default memory type.
shm

Use shared memory as the buffers. Allocated through shmget(2).
```
The option name has an empty line before the description text which is different to what's seen in http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#block-size . Additionally some formatting seems to have been misinterpreted:
```
Note that for shmhuge and m**maphuge** to 
```
(You can see the stars). Possibly the last nr_hugepages in "number in nr_hugepages" should be formatted.

http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-size hyperlink runtime, nrfiles and filesize?

More unexpected vertical alignment on http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#i-o-engine ? Also verbatim formatting for "direct=1" "buffered=0" "cpuload =85 " "/dev/dax0.0"?

In http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-rate_process the Wikipedia link redirects changed to https://en.wikipedia.org/wiki/Poisson_point_process  .

Errant quote in http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-flow :
"The ‘flow parameter"

Vertical spacing http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-verify .

What's "header_interval" in http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-verify_interval ? Should "blocksize" be hyperlinked?

http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-trim_backlog_batch : IO -> I/O?

Vertical alignment - http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-steadystate .

Perhaps the verbatim and regular text needs to be more interspersed in the example within http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#verification-trigger-example ...

Phew! That's an amazing conversion you've done!

termim commented:
Thank you for the review! See below.


On 01/15/17 04:28, Sitsofe Wheeler wrote:
>
> @termim <https://github.com/termim> Anything I find now is going to 
> minor - you've done a great job!
>
> (The buffer_pattern=0xdeadface problem has been resolved and looks as 
> I would expect by using non-smart quotes)
>
> Sometimes when writing "fio" the first letter capitalised at the 
> starts of sentences and sometimes it's not. Sometimes fio is written 
> in all caps. Sometimes fio is emboldened. For example:
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#source
> "fio resides in a git repo[...]"
>
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#author
> "Fio was written by Jens Axboe[...]"
>
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#building
> "To build FIO with a cross-compiler:"
>
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#platforms
> "Fio works on (at least) Linux, Solaris,[...]"
>
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption--section
> "Tell *fio* to run only the [...]"
>
> Personally like the capitalisation of the first letter at the start of 
> a sentence but otherwise I'd say fio should be all lowercase. Thoughts?
>
Agreed. Fixed.

> Sometimes paths have formatting sometimes they don't
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#windows
> "[...] dobuild.cmd from the os/windows"
> (|os/windows| is a path in this case and |dobuild.cmd| is actually a 
> filename)
>
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#documentation :
> "[...] and direct your browser to |./doc/output/html/index.html|."
>
Fixed.

> The dash in the following should probably be an em dash (were you 
> getting those by using -- ?):
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#how-fio-works :
> "A job file may contain any number of threads and/or files - the 
> typical contents of the job [...]"
>
Yes. Fixed.

> In 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#command-line-options 
> sometimes commands that have a singl sentence end in a full stop and 
> sometimes they don't:
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption--parse-only
> "Parse options only, don’t start any I/O"
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption--minimal
> "Print statistics in a terse, semicolon-delimited format."
>
Yes, a consistency would be good. What is the correct way?

> Perhaps now we have the ablility to use formatting the text in the 
> brackets should be changed in 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption--idle-prof 
> ? Other commands seem to use "", "" ...:
>
> |Report cpu idleness on a system or percpu basis 
> (option=`system,percpu`) or run unit work calibration only 
> (option=`calibrate`).|
|Fixed.

|
> ||
>
> Line wrapping looks strange and "af" should probably be "at" in 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption--trigger-timeout 
> :
> "*--trigger-timeout*=t Execute
> trigger af this time"
>
Fixed.

> In 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#job-file-format 
> one of the example job files is not formatted with .ini colouring 
> (making it stand out from the rest):
>
> |filename directive, as in the following example: ; -- start job file 
> including.fio -- [global] [...]|
|This is a deficiency of the pygments library used by Sphinx. It doesn't 
support flags (parameters without value) in INI files. It is already 
fixed in development of pygments but not released yet.

|
> ||
>
> Should there be extra formatting for things that are to do with 
> VARNAME in 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#environment-variables 
> ?
>
What about now?

> In 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#parameter-types 
> there is this:
>
> |1 MB: 1000000, 1m, 1000k 1 TiB: 1073741824, 1ti, 1024mi, 1048576ki 1 
> TB: 1000000000, 1t, 1000m, 1000000k |
>
> Isn't 1TiB = 1024**4 = 1099511627776 rather than 1073741824? Perhaps 
> the unit for that row were meant to be 1 GiB and 1gi? There might be 
> something similar later too:
>
> |1 MB: 1000000, 1mi, 1000ki 1 TiB: 1073741824, 1t, 1024m, 1048576k 1 
> TB: 1000000000, 1ti, 1000mi, 1000000ki|
I didn't look into this but now it seems not clear to me. Consider these 
two lines:

         Ki -- means kilo (K) or 1000

when kb_base=1000, and

         k -- means kibi (Ki) or 1024

when kb_base=1024, and

What actually Ki means? I'd expect that units name should not change 
it's meaning.
If user specified 1KiB it should mean 1024 whatever value kb_base has.

But it probably is not a question for the doc conversion.

> ||
>
> Perhaps there should be a back link to the int section of 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#parameter-types 
> in 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-kb_base 
> ?
>
I agree, but I couldn't find a way to do that yet :(

> More smart quotes in unexpected places in 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-directory 
> :
>
> |Used to place files in a different location than :file:”./”. |
|Fixed.

|
> ||
>
> In 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-file_service_type 
> perhaps the example should just be verbatim:
> "For example, specifying *file_service_type* = /random:8/ would"
> file_service_type is a self hyperlink but the whole thing is kind of 
> an example so I'd expect the same rendering as |rw=write:4k| in 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-readwrite 
> .
>
Fixed.

> Backslash needs escaping in 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-filename 
> :
> "[...] then escape that with a ‘’ character."
> (notice how there's nothing between the quotes)
>
Fixed.

> Missing formatting around "rw=randread:8" in 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-rw_sequencer 
> .
>
Fixed.

> The vertical spacing looks weird in 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-iomem :
>
> |malloc Use memory from malloc(3) as the buffers. Default memory type. 
> shm Use shared memory as the buffers. Allocated through shmget(2).|
|Fixed.

|
> ||
>
> The option name has an empty line before the description text which is 
> different to what's seen in 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#block-size . 
> Additionally some formatting seems to have been misinterpreted:
>
> |Note that for shmhuge and m**maphuge** to |
>
> (You can see the stars). Possibly the last nr_hugepages in "number in 
> nr_hugepages" should be formatted.
>
Fixed.

> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-size 
> hyperlink runtime, nrfiles and filesize?
>
Fixed.

> More unexpected vertical alignment on 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#i-o-engine ? Also 
> verbatim formatting for "direct=1" "buffered=0" "cpuload =85 " 
> "/dev/dax0.0"?
>
Fixed.

> In 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-rate_process 
> the Wikipedia link redirects changed to 
> https://en.wikipedia.org/wiki/Poisson_point_process .
>
Fixed.

> Errant quote in 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-flow :
> "The ‘flow parameter"
>
Fixed.

> Vertical spacing 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-verify 
> .
>
Fixed.

> What's "header_interval" in 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-verify_interval 
> ? Should "blocksize" be hyperlinked?
>
verify_interval probably. Fixed.

> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-trim_backlog_batch 
> : IO -> I/O?
>
Fixed.

> Vertical alignment - 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-steadystate 
>
>
Fixed.

> .
>
> Perhaps the verbatim and regular text needs to be more interspersed in 
> the example within 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#verification-trigger-example 
> ...
>
Fixed.

> Phew! That's an amazing conversion you've done!
>

Thanks! We are closer and closer :)

sitsofe commented:
@termim - great work!

> > In http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#command-line-options sometimes commands that have a singl sentence end in a full stop and sometimes they don't:
> > http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption--parse-only
> > "Parse options only, don’t start any I/O"
> > http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption--minimal
> > "Print statistics in a terse, semicolon-delimited format."
>
> Yes, a consistency would be good. What is the correct way?

I'd say always put a full stop. That way it won't matter if it's a multi-line sentence or not.

> > Should there be extra formatting for things that are to do with VARNAME in http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#environment-variables ?
>
> What about now?

Looks good.

> >  In
> >  http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#parameter-types
> >  there is this:
> >
> >  |1 MB: 1000000, 1m, 1000k 1 TiB: 1073741824, 1ti, 1024mi, 1048576ki 1
> >  TB: 1000000000, 1t, 1000m, 1000000k |
> >
> >  Isn't 1TiB = 1024**4 = 1099511627776 rather than 1073741824? Perhaps
> >  the unit for that row were meant to be 1 GiB and 1gi? There might be
> >  something similar later too:
> >
> >  |1 MB: 1000000, 1mi, 1000ki 1 TiB: 1073741824, 1t, 1024m, 1048576k 1
> >  TB: 1000000000, 1ti, 1000mi, 1000000ki|
>
> I didn't look into this but now it seems not clear to me. Consider these
> two lines:
>
>          Ki -- means kilo (K) or 1000
>
> when kb_base=1000, and
>
>          k -- means kibi (Ki) or 1024
>
> when kb_base=1024, and
>
> What actually Ki means? I'd expect that units name should not change
> it's meaning.
> If user specified 1KiB it should mean 1024 whatever value kb_base has.
>
> But it probably is not a question for the doc conversion.

The confusion around the reversal of meaning for things like KiB in inputs is described here:
http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-kb_base
"Compatibility mode (default). To avoid breaking old scripts:

power-of-2 values with SI prefixes
power-of-10 values with IEC prefixes"

In an ideal world KiB would have always meant "power of two" but it didn't happen that way (you can see the complaints over on https://github.com/axboe/fio/issues/103 ). My question above was more about the magnitude of the value rather than the base:

"1 TiB: 1073741824, 1ti, 1024mi, 1048576ki"

Assuming powers of 2, a TiB is 1024**4 which is a magnitude bigger than 1073741824 == 1024 **3 and thus is also a magnitude bigger than 1024mi or 1048576ki on the same line. So I wonder whether 1 TiB was meant to be 1 GiB and 1ti was meant to be 1gi in that line...

Something I didn't spot previously - http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-filename is missing backslashes on the Windows paths. For example `\.PhysicalDrive0` should be `\\.\PhysicalDrive0`.

I wonder if the filename hyperlinks in http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-filename should be undone (we're in that section after all) and made verbatim with the following text e.g. it shows "filename = /dev/sda:/dev/sdb" at the moment..

sitsofe commented:
One more: http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#i-o-engine under posixaio it looks like some old formatting made it into the prose itself: "Baio_readfR|(3) and aio_write(3)" (I'm guessing the B and R shouldn't be present).

termim commented:
On 01/16/17 02:40, Sitsofe Wheeler wrote:

> @termim <https://github.com/termim> - great work!
>
>         In
>         http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#command-line-options
>         sometimes commands that have a singl sentence end in a full
>         stop and sometimes they don't:
>         http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption--parse-only
>         "Parse options only, don’t start any I/O"
>         http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption--minimal
>         "Print statistics in a terse, semicolon-delimited format."
>
>     Yes, a consistency would be good. What is the correct way?
>
> I'd say always put a full stop. That way it won't matter if it's a 
> multi-line sentence or not.
>
Done.

>         Should there be extra formatting for things that are to do
>         with VARNAME in
>         http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#environment-variables
>         ?
>
>     What about now?
>
> Looks good.
>
>         In
>         http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#parameter-types
>         there is this:
>
>         |1 MB: 1000000, 1m, 1000k 1 TiB: 1073741824, 1ti, 1024mi,
>         1048576ki 1
>         TB: 1000000000, 1t, 1000m, 1000000k |
>
>         Isn't 1TiB = 1024**4 = 1099511627776 rather than 1073741824?
>         Perhaps
>         the unit for that row were meant to be 1 GiB and 1gi? There
>         might be
>         something similar later too:
>
>         |1 MB: 1000000, 1mi, 1000ki 1 TiB: 1073741824, 1t, 1024m,
>         1048576k 1
>         TB: 1000000000, 1ti, 1000mi, 1000000ki|
>
>     I didn't look into this but now it seems not clear to me. Consider
>     these
>     two lines:
>
>     |Ki -- means kilo (K) or 1000 |
>
>     when kb_base=1000, and
>
>     |k -- means kibi (Ki) or 1024 |
>
>     when kb_base=1024, and
>
>     What actually Ki means? I'd expect that units name should not change
>     it's meaning.
>     If user specified 1KiB it should mean 1024 whatever value kb_base has.
>
>     But it probably is not a question for the doc conversion.
>
> The confusion around the reversal of meaning for things like KiB in 
> inputs is described here:
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-kb_base
> "Compatibility mode (default). To avoid breaking old scripts:
>
> power-of-2 values with SI prefixes
> power-of-10 values with IEC prefixes"
>
> In an ideal world KiB would have always meant "power of two" but it 
> didn't happen that way (you can see the complaints over on #103 
> <https://github.com/axboe/fio/issues/103> ). My question above was 
> more about the magnitude of the value rather than the base:
>
> "1 TiB: 1073741824, 1ti, 1024mi, 1048576ki"
>
> Assuming powers of 2, a TiB is 1024**4 which is a magnitude bigger 
> than 1073741824 == 1024 **3 and thus is also a magnitude bigger than 
> 1024mi or 1048576ki on the same line. So I wonder whether 1 TiB was 
> meant to be 1 GiB and 1ti was meant to be 1gi in that line...
>
Done.

> Something I didn't spot previously - 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-filename 
> is missing backslashes on the Windows paths. For example 
> |\.PhysicalDrive0| should be |\\.\PhysicalDrive0|.
>
Done.

> I wonder if the filename hyperlinks in 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-filename 
> should be undone (we're in that section after all) and made verbatim 
> with the following text e.g. it shows "filename = /dev/sda:/dev/sdb" 
> at the moment..
>
Done.

termim commented:
On 01/16/17 07:30, Sitsofe Wheeler wrote:
>
> One more: 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#i-o-engine under 
> posixaio it looks like some old formatting made it into the prose 
> itself: "Baio_readfR|(3) and aio_write(3)" (I'm guessing the B and R 
> shouldn't be present).
>
Done.

sitsofe commented:
Oops I made you over escape. In http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-filename 'filename="/dev/dsk/foo@3,0\\:c"' only needs to be 'filename="/dev/dsk/foo@3,0\:c"' (that path looks like a Solaris disk not a Windows one).

http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#parameter-types
"1 TiB: 1099511627776, 1ti, 1024mi, 1048576ki"

still looks wrong. 1 TiB is 1099511627776 is 1ti so that's part is correct but 1024mi and 1048576ki  would only be 1073741824 which is a magnitude smaller. So 1024mi -> 1024gi, 1048576ki -> 1048576mi ?

@axboe What do you think about moving the HOWTO and README files to have .rst extensions so they are rendered more nicely in github?

sitsofe commented:
In case it wasn't clear (because of how things will appear in email and on github web pages) the previous comment was asking for only one backslash to appear in the '/dev/dsk/foo@3,0...' example because it is demonstrating the need to escape a colon (as opposed to escaping a backslash like the Windows examples that were fixed).

sitsofe commented:
Looking at http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html where does the version number (fio - Flexible I/O tester rev. 2.15) in the heading come from? Should it be 2.16?

axboe commented:
Thanks for taking the time to look into this as well, Sitsofe. I cut a 2.17 release this morning, and was planning on getting this in at the start of the new cycle. Can you guys let me know when you think it's ready for merging? Then I'll take a look.

sitsofe commented:
@axboe I think it's pretty much ready and @termim has worked tirelessly to keep the patches up to date with respect to other commits that have gone in to fio so I'd love to see this go in before another commit goes and breaks it further.

I'll just repeat my previous question in case it was missed. Jens: What do you think about moving the HOWTO and README files to have .rst extensions so they are rendered more nicely on github web pages (see https://github.com/termim/fio/blob/sphinx-doc/HOWTO for an example of how it is rendered currently)?

axboe commented:
I'm fine with getting it in sooner rather than later. Unfortunately, there's a current conflict that would be great it could be resolved on that end before I pull it in.

I'm fine with .rst extensions for README and HOWTO.

termim commented:
On 01/17/17 02:16, Sitsofe Wheeler wrote:

> Oops I made you over escape. In 
> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#cmdoption-arg-filename 
> 'filename="/dev/dsk/foo@3,0\:c"' only needs to be 
> 'filename="/dev/dsk/foo@3,0:c"' (that path looks like a Solaris disk 
> not a Windows one).
>
Done.

> http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#parameter-types
> "1 TiB: 1099511627776, 1ti, 1024mi, 1048576ki"
>
> still looks wrong. 1 TiB is 1099511627776 is 1ti so that's part is 
> correct but 1024mi and 1048576ki would only be 1073741824 which is a 
> magnitude smaller. So 1024mi -> 1024gi, 1048576ki -> 1048576mi ?
>
Do you mean 1024m -> 1024g, 1048576k -> 1048576mas well?

termim commented:
On 01/17/17 04:23, Sitsofe Wheeler wrote:

> Looking at http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html where 
> does the version number (fio - Flexible I/O tester rev. 2.15) in the 
> heading come from? Should it be 2.16?
>
It comes from FIO-VERSION-FILE. see conf.py, fio_version() function.

termim commented:
On 01/17/17 17:12, Jens Axboe wrote:

> I'm fine with getting it in sooner rather than later. Unfortunately, 
> there's a current conflict that would be great it could be resolved on 
> that end before I pull it in.
>
> I'm fine with .rst extensions for README and HOWTO.
>

I've rebased to the latest master, fixed last comments from Sitsofe and 
renamed README and HOWTO as requested. I think you can do the final 
review now.

termim commented:
On Wed, Jan 18, 2017 at 2:43 AM, Sitsofe Wheeler <notifications@github.com>
wrote:

> Renaming those files to have a .rst extension didn't go well because most
> restructuredtext parsers can't handle Sphinx's extra directives (such as
> :option: - http://www.sphinx-doc.org/en/1.5.1/markup/inline.html?
> highlight=manpage#role-option ) and just display errors or nothing. For
> example, see how "Job description" section is empty on
> https://gist.github.com/sitsofe/5610dafcddc5c89c8dbb9e5d3bce66
> 8d#job-description , sigh. @termim <https://github.com/termim> I'm really
> sorry to ask but could you undo the file renaming?

No problem. Done

sitsofe commented:
@termim OK I'm happy (at least for now) - would you say it's ready for Jens to merge?

termim commented:
On Wed, Jan 18, 2017 at 10:44 AM, Sitsofe Wheeler <notifications@github.com>
wrote:

> @termim <https://github.com/termim> OK I'm happy (at least for now) -
> would you say it's ready for Jens to merge?
>
Yes.

axboe commented:
I merged this, thanks for all your work. However, I still think there's work left. For instance, the fio.1 should now be auto-generated. Not as part of the regular build, but by me whenever we modify the documentation. I'll typically do that for a release, for instance.

Is there any pending work that needs to be done before I can git rm fio.1?

sitsofe commented:
Jens: are you saying there's a release target which should (re)generate things for tarball preparation  or are you wondering whether fio.1 should be deleted?

axboe commented:
Doesn't have to, I'm fine with running it manually. If/when I get tired of that, I'll fix that up. I don't want to carry around duplicate information is my main issue. One huge pain previously has been that new options or changes to options would need to be documented in both HOWTO and fio.1, which causes drift between them. That's the main issue I'm looking to solve with this sphinx change.

termim commented:
On Thu, Jan 19, 2017 at 2:41 PM, Jens Axboe <notifications@github.com>
wrote:

> I merged this, thanks for all your work. However, I still think there's
> work left. For instance, the fio.1 should now be auto-generated. Not as
> part of the regular build, but by me whenever we modify the documentation.
> I'll typically do that for a release, for instance.
>
> Is there any pending work that needs to be done before I can git rm fio.1?
>
It is already there, just run `make -C doc man` and man page will be in
doc/output/man/fio.1
See http://fio.readthedocs.io/en/sphinx-doc/fio_doc.html#documentation

sitsofe commented:
@termim but it's safe for https://github.com/axboe/fio/blob/master/fio.1 to be removed now?

termim commented:
On Thu, Jan 19, 2017 at 4:06 PM, Sitsofe Wheeler <notifications@github.com>
wrote:

> @termim <https://github.com/termim> but it's safe for
> https://github.com/axboe/fio/blob/master/fio.1 to be removed now?
>
I've synced the job file options in HOQTO and old fio.1. Probably it would
be good to check other sections as well. The formatting of generated fio.1
is not as good as formatting of the manually written fio.1 though -
sometimes vertical spacing is too big IMHO, but I wouldn't call it a show
stopper.

sitsofe commented:
I've noticed a few warning pages while using the generated man page:

```
<standard input>:2004: warning [p 21, 2.2i, div `an-div', 0.0i]: cannot adjust line
<standard input>:2004: warning [p 21, 2.2i]: cannot adjust line
<standard input>:3874: warning [p 37, 7.2i, div `3tbd4,2', 0.2i]: cannot adjust line
grotty:<standard input> (<standard input>):41654: character above first line discarded
grotty:<standard input> (<standard input>):41658: character above first line discarded
grotty:<standard input> (<standard input>):41662: character above first line discarded
grotty:<standard input> (<standard input>):41666: character above first line discarded
```

Also some words suffer end of line hyphenation when  I wouldn't have expected it:
```
       --idle-prof=option
              Report cpu idleness on a system or percpu basis --idle-prof=sys-
              tem,percpu  or  run unit work calibration only --idle-prof=cali-
              brate.
```

I'm guessing there's not much to be done about these until Sphinx fixes it.

termim commented:
The heading for the "Interpreting the output" section is missing in generated fio.1 ...
I've filed an issue to the Sphinx https://github.com/sphinx-doc/sphinx/issues/3360

axboe commented:
It's not the only section missing. Just checked a few after that, and most of them are not there... The section is there, but the header/title is not.

sitsofe commented:
The problem is that the first level headings are ignored when using Sphinx's man page conversion. Looking at styles:


```
Chapter 1 Title
===============

Section 1.2 Title
-----------------

Subsection 1.2.3 Title
~~~~~~~~~~~~~~~~~~~~~~

Section 2.2 Title
-----------------

Chapter 3 Title
===============
```

Chapter 1 Title and Chapter 3 Title won't appear in the man output (note that ReStructuredText works out the heading level on the fly so those are treated as top level headings)
```
Section 1 Title
---------------

Subsection 1.2 Title
~~~~~~~~~~~~~~~~~~~~

Section 2 Title
---------------
```

Only Subsection 1.2 will be printed in the man page output.

Poking at the Sphinx source it sort of looks like it's by design: https://github.com/sphinx-doc/sphinx/blob/1.5.1/sphinx/writers/manpage.py#L410 .

sitsofe commented:
Does the pull request https://github.com/axboe/fio/pull/290 help?

termim commented:
Jens, while converting to ReST I've created a project on readthedocs.org, see
http://fio.readthedocs.io/en/latest/
to test the conversion. Now when technical conversion is done, do you want to take over the readthedocs project or create your own one? In the later case I'd need to close mine to keep the links nice. IMHO it is logical to have the project itself and the docs under the same umbrella.

sitsofe commented:
@termim perhaps we should take this to the mailing list. I find fio.readthedocs.io to be useful for reading/linking to because all the formatting codes have been applied...
